Add import for classes in qiskit/quantum_info/synthesis

ShellyGarion opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fix a problem that was mentioned here:
https://github.com/Qiskit/qiskit/pull/11635#issuecomment-1915526756

This line:
`from qiskit.quantum_info.synthesis.two_qubit_decompose import TwoQubitWeylDecomposition`
will work in Qiskit 0.46 and raise a deprecation warning.

### Details and comments

ShellyGarion committed:
add import for TwoQubitWeylDecomposition

ShellyGarion committed:
fix lint errors

mtreinish commented on qiskit/quantum_info/synthesis/two_qubit_decompose.py:
```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+import warnings
+
+from qiskit.synthesis.two_qubit.two_qubit_decompose import TwoQubitWeylDecomposition
```
There are a few public decomposer classes in the module. It's probably better to play it safe with:

```suggestion
from qiskit.synthesis.two_qubit.two_qubit_decompose import *
```

and expose them all.

ShellyGarion replied:
When I try to do this change, I get the following pylint errors:


```
qiskit/quantum_info/synthesis/two_qubit_decompose.py:30:0: W0401: Wildcard import qiskit.synthesis.two_qubit.two_qubit_decompose (wildcard-import)
qiskit/quantum_info/synthesis/two_qubit_decompose.py:30:0: W0614: Unused import(s) cmath, math, io, base64, logging, np, logger, decompose_two_qubit_product_gate, TwoQubitWeylDecomposition, TwoQubitWeylIdEquiv, TwoQubitWeylSWAPEquiv, TwoQubitWeylPartialSWAPEquiv, TwoQubitWeylPartialSWAPFlipEquiv, TwoQubitWeylControlledEquiv, TwoQubitControlledUDecomposer, TwoQubitWeylMirrorControlledEquiv, TwoQubitWeylfSimaabEquiv, TwoQubitWeylfSimabbEquiv, TwoQubitWeylfSimabmbEquiv, TwoQubitWeylGeneral, Ud, trace_to_fid, rz_array, TwoQubitBasisDecomposer, TwoQubitDecomposeUpToDiagonal, two_qubit_cnot_decompose, ClassVar, Optional, Type, QuantumRegister, QuantumCircuit, Gate, CXGate, RXGate, RYGate, RZGate, QiskitError, Operator, weyl_coordinates, transform_to_magic_basis, OneQubitEulerDecomposer, DEFAULT_ATOL and deprecate_arg from wildcard import of qiskit.synthesis.two_qubit.two_qubit_decompose (unused-wildcard-import)

```

ShellyGarion replied:
I can add more classes to the import (and later also to the main branch documentation), but I think it would be better to state them specifically

ShellyGarion replied:
Note that except of `TwoQubitWeylDecomposition, TwoQubitControlledUDecomposer, TwoQubitDecomposeUpToDiagonal`, all other classes are variants of `TwoQubitWeylDecomposition`.

ShellyGarion replied:
A search that I did in various libraries (`circuit knitting`, `qiskit-ibm-runtime` and `qiskit-ibm-provider`) does not indicate the usage of other classes except of `TwoQubitWeylDecomposition`.

mtreinish replied:
I was mostly worried about other downstream users that we don't know about, or that don't have public code anywhere that we can access. I think listing all the public classes explicitly is fine, being explicit doesn't hurt anything. It'll fix the pylint errors too.

ShellyGarion replied:
OK, I added all the classes in the file `two_qubit_decompose`

Cryoris replied:
> I was mostly worried about other downstream users that we don't know about

Though there's loads of other specific imports that are still going to break. Here we're only fixing the two-qubit one, but before (<=0.45) people were able to import e.g. `qiskit.quantum_info.synthesis.one_qubit_decompose`, `.qsd`, `.weyl`, `.local_invariance`....

mtreinish replied:
That's a fair point, on the one hand we're documenting this release as "fully compatible with the Qiskit 0.45.x releases" but with extra deprecations. But we've also documented the public interface as those that are documented. This is a weird case because `qiskit.quantum_info.synthesis` is a dark corner of the API where there was a ton of undocumented functionality that wasn't exposed anywhere except in the leaf modules, and the definition of the stable api only really changed after 0.45 (in preparation for 1.0.0).

I'm inclined to just err on the side of caution more here just so people don't have an excuse for not upgrading to 0.46 and add module shims back in for everything that redirects to the new location in`qiskit.synthesis` with a deprecation warning. (like what's done for `__init__.py` here already)

mtreinish reviewed (commented):
I think this is the right call, just one inline suggestion. Otherwise LGTM

I think there is likely some follow up work on `main` to expose the various specialized two qubit decomposer classes as public and include them in the documentation.

ShellyGarion committed:
add more classes

Cryoris commented on qiskit/quantum_info/synthesis/two_qubit_decompose.py:
```diff
@@ -0,0 +1,51 @@
+import warnings
+
+from qiskit.synthesis.two_qubit.two_qubit_decompose import (
+    TwoQubitWeylDecomposition,
```
It seems to be missing the `TwoQubitBasisDecomposer`, or is there a reason it's not here? 🙂

ShellyGarion replied:
Yes, there is a reason. This class already appeared in qiskit/quanutm_info init file, and was properly deprecated in #11460.
This PRs handles classes that have never properly appeared in qiskit/quanutm_info init file before.

Cryoris commented on qiskit/quantum_info/synthesis/two_qubit_decompose.py:
```diff
@@ -0,0 +1,51 @@
+from __future__ import annotations
+import warnings
+
+from qiskit.synthesis.two_qubit.two_qubit_decompose import (
```
While we're at it, could we then also add the `decompose_two_qubit_product_gate`? It also has a docstring and might have been mistaken as "public"

ShellyGarion replied:
I think that this function is used internally in some of the classes.

ShellyGarion commented:
https://github.com/Qiskit/qiskit/pull/11672#discussion_r1472774826 - I think that these functions are used internally (in 2q unitary or general unitary synthesis)

mtreinish commented on qiskit/quantum_info/synthesis/two_qubit_decompose.py:
```diff
@@ -0,0 +1,51 @@
+    TwoQubitWeylfSimabbEquiv,
+    TwoQubitWeylfSimabmbEquiv,
+    TwoQubitWeylGeneral,
+)
```
After thinking it over some more I feel a wildcard import is the proper thing to do here. We can just tell pylint to ignore the rules it was failing before:

```suggestion
# pylint: disable=wildcard-import,unused-wildcard-import

from qiskit.synthesis.two_qubit.two_qubit_decompose import *
```

mtreinish reviewed (commented):
I think @Cryoris is correct that we'll want similar redirect shims for the other leaf modules that were there before. In the interest of time (since we're scheduled to release tomorrow: 2024-02-01) I can just add the extra modules in #11697 after this PR merges.

ShellyGarion committed:
add back all imports

ShellyGarion commented:
> I think @Cryoris is correct that we'll want similar redirect shims for the other leaf modules that were there before. In the interest of time (since we're scheduled to release tomorrow: 2024-02-01) I can just add the extra modules in #11697 after this PR merges.

@mtreinish, @Cryoris:
I did as you suggested. I added `import *` from the files: `two_qubit_decompose`, `local_invariance`, `weyl` and `qsd`.
There is no need to handle `one_qubit_decompose` since it contains only one class which was properly documented (and no other classes or functions). 
I think this is a better approach than doing it in #11697.

ShellyGarion committed:
also add qsd

ShellyGarion commented:
I think this PR handles all files that were originally under `qiskit.qauntum_info.synthesis`:
https://github.com/Qiskit/qiskit/tree/stable/0.45/qiskit/quantum_info/synthesis

- clifford_decompose - deprecated in a previous release
- cnotdihedral_decompose - deprecated in a previous release
- ion_decmpose - was deprecated in #11460, and removed for qiskit 1.0
- local_invariance - added here, and deprecated
- one_qubit_decompose - was deprecated in #11460, and moved to qiskit.synthesis
- qsd - added here, and deprecated
- quaternion - moved to qiskit.quantum_info in #11460 
- one_qubit_decompose - added here, and deprecated
- weyl - added here, and deprecated
- xx_decompose - was deprecated in #11460, and moved to qiskit.synthesis

ShellyGarion committed:
style

mtreinish reviewed (approved):
This LGTM, thanks for doing this @ShellyGarion
